In [1]:
# Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

import warnings
warnings.simplefilter('ignore')

### 1. Loading data

In [4]:
# Read data
prev_apps = pd.read_csv('previous_application.csv')

In [5]:
prev_apps.head()

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START  ... NAME_SELLER_INDUSTRY  CNT_PAYMENT  \
0                       15  ...         Connectivity         12.0   
1                       11  ...                  XNA         36.0   
2                       11  ...                  XNA         12.0   
3                        7  ...                  XNA         12.0   
4                        9  ...                  XNA         24.0   

   NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0            middle  POS mobile with interest            365243.0   
1        low_action          Cash X-Sell: low            365243.0   
2              high         Cash X-Sell: high            365243.0   
3            middle       Cash X-Sell: middle            365243.0   
4              high         Cash Street: high                 NaN   

  DAYS_FIRST_DUE DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
0          -42.0                     300.0          -42.0            -37.0   
1         -134.0                     916.0       365243.0         365243.0   
2         -271.0                      59.0       365243.0         365243.0   
3         -482.0                    -152.0         -182.0           -177.0   
4            NaN                       NaN            NaN              NaN   

  NFLAG_INSURED_ON_APPROVAL  
0                       0.0  
1                       1.0  
2                       1.0  
3                       1.0  
4                       NaN  

[5 rows x 37 columns]

In [15]:
prev_apps.shape

(1670214, 37)

In [3]:
cur_apps = pd.read_csv('application_data.csv')

In [6]:
cur_apps.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 122 columns]

In [14]:
cur_apps.shape

(307511, 122)

### 2. Merge data into single dataframe

In [7]:
all_apps = cur_apps.merge(prev_apps, how='left', on='SK_ID_CURR')

In [8]:
all_apps.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE_x CODE_GENDER FLAG_OWN_CAR  \
0      100002       1           Cash loans           M            N   
1      100003       0           Cash loans           F            N   
2      100003       0           Cash loans           F            N   
3      100003       0           Cash loans           F            N   
4      100004       0      Revolving loans           M            Y   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT_x  \
0               Y             0          202500.0      406597.5   
1               N             0          270000.0     1293502.5   
2               N             0          270000.0     1293502.5   
3               N             0          270000.0     1293502.5   
4               Y             0           67500.0      135000.0   

   AMT_ANNUITY_x  ...  NAME_SELLER_INDUSTRY CNT_PAYMENT NAME_YIELD_GROUP  \
0        24700.5  ...       Auto technology        24.0       low_normal   
1        35698.5  ...                   XNA        12.0       low_normal   
2        35698.5  ...             Furniture         6.0           middle   
3        35698.5  ...  Consumer electronics        12.0           middle   
4         6750.0  ...          Connectivity         4.0           middle   

           PRODUCT_COMBINATION DAYS_FIRST_DRAWING DAYS_FIRST_DUE  \
0      POS other with interest           365243.0         -565.0   
1             Cash X-Sell: low           365243.0         -716.0   
2   POS industry with interest           365243.0         -797.0   
3  POS household with interest           365243.0        -2310.0   
4  POS mobile without interest           365243.0         -784.0   

   DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE  DAYS_TERMINATION  \
0                      125.0          -25.0             -17.0   
1                     -386.0         -536.0            -527.0   
2                     -647.0         -647.0            -639.0   
3                    -1980.0        -1980.0           -1976.0   
4                     -694.0         -724.0            -714.0   

   NFLAG_INSURED_ON_APPROVAL  
0                        0.0  
1                        1.0  
2                        0.0  
3                        1.0  
4                        0.0  

[5 rows x 158 columns]

In [13]:
all_apps.shape

(1430155, 158)

### 3. Missing values

In [16]:
all_apps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1430155 entries, 0 to 1430154
Columns: 158 entries, SK_ID_CURR to NFLAG_INSURED_ON_APPROVAL
dtypes: float64(85), int64(41), object(32)
memory usage: 1.7+ GB


In [17]:
all_apps.columns

Index(['SK_ID_CURR', 'TARGET', 'NAME_CONTRACT_TYPE_x', 'CODE_GENDER',
       'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN', 'AMT_INCOME_TOTAL',
       'AMT_CREDIT_x', 'AMT_ANNUITY_x',
       ...
       'NAME_SELLER_INDUSTRY', 'CNT_PAYMENT', 'NAME_YIELD_GROUP',
       'PRODUCT_COMBINATION', 'DAYS_FIRST_DRAWING', 'DAYS_FIRST_DUE',
       'DAYS_LAST_DUE_1ST_VERSION', 'DAYS_LAST_DUE', 'DAYS_TERMINATION',
       'NFLAG_INSURED_ON_APPROVAL'],
      dtype='object', length=158)

In [31]:
# Store variables
cols = all_apps.columns
sz = all_apps.shape[0]

In [33]:
na_df = pd.DataFrame(columns=['name', 'null count', 'null percent', 'type'])

for col in cols:
    na_sum = all_apps[col].isna().sum()
    if na_sum != 0:
        na_percent = na_sum / sz * 100
        na_df = na_df.append({'name': col, 'null count': na_sum, 'type': all_apps[col].dtype, 'null percent': na_percent}, ignore_index=True)



In [34]:
na_df.sort_values(by='null percent', ascending=False, inplace=True)

In [35]:
na_df

name null count  null percent     type
27           COMMONAREA_MODE     995502     69.607980  float64
13            COMMONAREA_AVG     995502     69.607980  float64
41           COMMONAREA_MEDI     995502     69.607980  float64
35  NONLIVINGAPARTMENTS_MODE     989835     69.211729  float64
21   NONLIVINGAPARTMENTS_AVG     989835     69.211729  float64
..                       ...        ...           ...      ...
7               EXT_SOURCE_2       1901      0.132923  float64
1          AMT_GOODS_PRICE_x       1219      0.085236  float64
0              AMT_ANNUITY_x         93      0.006503  float64
5            CNT_FAM_MEMBERS          2      0.000140  float64
60    DAYS_LAST_PHONE_CHANGE          1      0.000070  float64

[99 rows x 4 columns]

#### 3.1. Dropping columns with > 50% values missing

In [45]:
large_empty_cols = na_df[na_df['null percent'] > 50]['name'].values

In [46]:
large_empty_cols.size

34

In [47]:
type(large_empty_cols)

numpy.ndarray

In [48]:
all_apps.drop(large_empty_cols, axis=1, inplace=True)

In [50]:
# Verifying that the correct number of columns were dropped
# There were 154 before, and 34 were dropped, leaving 120.
all_apps.shape

(1430155, 120)

### 3.2. Dropping rows for columns with < 2% values missing

In [78]:
small_empty_cols = na_df[na_df['null percent'] < 2]['name'].values

In [79]:
small_empty_cols

array(['PRODUCT_COMBINATION', 'AMT_CREDIT_y', 'NAME_YIELD_GROUP',
       'NAME_CLIENT_TYPE', 'NAME_GOODS_CATEGORY',
       'NAME_CASH_LOAN_PURPOSE', 'CHANNEL_TYPE', 'NAME_PORTFOLIO',
       'NAME_PRODUCT_TYPE', 'NAME_PAYMENT_TYPE', 'SELLERPLACE_AREA',
       'CODE_REJECT_REASON', 'WEEKDAY_APPR_PROCESS_START_y',
       'DAYS_DECISION', 'NAME_CONTRACT_STATUS', 'NFLAG_LAST_APPL_IN_DAY',
       'FLAG_LAST_APPL_PER_CONTRACT', 'HOUR_APPR_PROCESS_START_y',
       'AMT_APPLICATION', 'NAME_CONTRACT_TYPE_y', 'SK_ID_PREV',
       'NAME_SELLER_INDUSTRY', 'NAME_TYPE_SUITE_x',
       'OBS_30_CNT_SOCIAL_CIRCLE', 'DEF_30_CNT_SOCIAL_CIRCLE',
       'DEF_60_CNT_SOCIAL_CIRCLE', 'OBS_60_CNT_SOCIAL_CIRCLE',
       'EXT_SOURCE_2', 'AMT_GOODS_PRICE_x', 'AMT_ANNUITY_x',
       'CNT_FAM_MEMBERS', 'DAYS_LAST_PHONE_CHANGE'], dtype=object)

In [80]:
na_df[na_df['null percent'] < 2]

name null count  null percent     type
92           PRODUCT_COMBINATION      16767      1.172390   object
71                  AMT_CREDIT_y      16455      1.150575  float64
91              NAME_YIELD_GROUP      16454      1.150505   object
83              NAME_CLIENT_TYPE      16454      1.150505   object
84           NAME_GOODS_CATEGORY      16454      1.150505   object
77        NAME_CASH_LOAN_PURPOSE      16454      1.150505   object
87                  CHANNEL_TYPE      16454      1.150505   object
85                NAME_PORTFOLIO      16454      1.150505   object
86             NAME_PRODUCT_TYPE      16454      1.150505   object
80             NAME_PAYMENT_TYPE      16454      1.150505   object
88              SELLERPLACE_AREA      16454      1.150505  float64
81            CODE_REJECT_REASON      16454      1.150505   object
73  WEEKDAY_APPR_PROCESS_START_y      16454      1.150505   object
79                 DAYS_DECISION      16454      1.150505  float64
78          NAME_CONTRACT_STATUS      16454      1.150505   object
76        NFLAG_LAST_APPL_IN_DAY      16454      1.150505  float64
75   FLAG_LAST_APPL_PER_CONTRACT      16454      1.150505   object
74     HOUR_APPR_PROCESS_START_y      16454      1.150505  float64
70               AMT_APPLICATION      16454      1.150505  float64
68          NAME_CONTRACT_TYPE_y      16454      1.150505   object
67                    SK_ID_PREV      16454      1.150505  float64
89          NAME_SELLER_INDUSTRY      16454      1.150505   object
2              NAME_TYPE_SUITE_x       4031      0.281858   object
56      OBS_30_CNT_SOCIAL_CIRCLE       3393      0.237247  float64
57      DEF_30_CNT_SOCIAL_CIRCLE       3393      0.237247  float64
59      DEF_60_CNT_SOCIAL_CIRCLE       3393      0.237247  float64
58      OBS_60_CNT_SOCIAL_CIRCLE       3393      0.237247  float64
7                   EXT_SOURCE_2       1901      0.132923  float64
1              AMT_GOODS_PRICE_x       1219      0.085236  float64
0                  AMT_ANNUITY_x         93      0.006503  float64
5                CNT_FAM_MEMBERS          2      0.000140  float64
60        DAYS_LAST_PHONE_CHANGE          1      0.000070  float64

In [85]:
all_apps[all_apps['PRODUCT_COMBINATION'].isna()]

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE_x CODE_GENDER FLAG_OWN_CAR  \
68           100024       0      Revolving loans           M            Y   
97           100031       1           Cash loans           F            N   
201          100058       0      Revolving loans           F            N   
285          100076       0           Cash loans           M            Y   
394          100096       1           Cash loans           F            N   
...             ...     ...                  ...         ...          ...   
1428811      455951       0           Cash loans           M            Y   
1429508      456113       0           Cash loans           M            N   
1429720      456155       0           Cash loans           F            N   
1429859      456187       0           Cash loans           M            Y   
1430031      456229       0      Revolving loans           F            N   

        FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT_x  \
68                    Y             0          135000.0      427500.0   
97                    Y             0          112500.0      979992.0   
201                   Y             0           54000.0      135000.0   
285                   N             0          180000.0      315000.0   
394                   Y             0           81000.0      252000.0   
...                 ...           ...               ...           ...   
1428811               N             2          247500.0     1125000.0   
1429508               Y             3          180000.0     1096020.0   
1429720               Y             0          202500.0     1971072.0   
1429859               Y             1          315000.0     1175314.5   
1430031               Y             1          270000.0      180000.0   

         AMT_ANNUITY_x  ...  NAME_SELLER_INDUSTRY CNT_PAYMENT  \
68             21375.0  ...                   NaN         NaN   
97             27076.5  ...                   NaN         NaN   
201             6750.0  ...                   NaN         NaN   
285             9679.5  ...                   NaN         NaN   
394            14593.5  ...                   NaN         NaN   
...                ...  ...                   ...         ...   
1428811        50314.5  ...                   NaN         NaN   
1429508        56092.5  ...                   NaN         NaN   
1429720        68643.0  ...                   NaN         NaN   
1429859        38974.5  ...                   NaN         NaN   
1430031         9000.0  ...                   NaN         NaN   

        NAME_YIELD_GROUP PRODUCT_COMBINATION DAYS_FIRST_DRAWING  \
68                   NaN                 NaN                NaN   
97                   NaN                 NaN                NaN   
201                  NaN                 NaN                NaN   
285                  NaN                 NaN                NaN   
394                  NaN                 NaN                NaN   
...                  ...                 ...                ...   
1428811              NaN                 NaN                NaN   
1429508              NaN                 NaN                NaN   
1429720              NaN                 NaN                NaN   
1429859              NaN                 NaN                NaN   
1430031              NaN                 NaN                NaN   

        DAYS_FIRST_DUE  DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE  \
68                 NaN                        NaN            NaN   
97                 NaN                        NaN            NaN   
201                NaN                        NaN            NaN   
285                NaN                        NaN            NaN   
394                NaN                        NaN            NaN   
...                ...                        ...            ...   
1428811            NaN                        NaN            NaN   
1429508            NaN                        NaN            NaN   
1429720  

In [86]:
all_apps.shape

(1430155, 120)

In [87]:
# Drop using loop
for col in small_empty_cols:
    all_apps = all_apps[~all_apps[col].isna()]
    

In [88]:
#Verification
all_apps.shape

(1404812, 120)

### 3.3. Checking remaining columns with missing values

In [89]:
med_empty_cols = na_df[(na_df['null percent'] >= 2) & (na_df['null percent'] <= 50)]

In [90]:
med_empty_cols

name null count  null percent     type
43                ENTRANCES_MEDI     713388     49.881866  float64
29                ENTRANCES_MODE     713388     49.881866  float64
15                 ENTRANCES_AVG     713388     49.881866  float64
82             NAME_TYPE_SUITE_y     711126     49.723701   object
34               LIVINGAREA_MODE     711001     49.714961  float64
48               LIVINGAREA_MEDI     711001     49.714961  float64
20                LIVINGAREA_AVG     711001     49.714961  float64
52                HOUSETYPE_MODE     710772     49.698949   object
44                FLOORSMAX_MEDI     705019     49.296685  float64
30                FLOORSMAX_MODE     705019     49.296685  float64
16                 FLOORSMAX_AVG     705019     49.296685  float64
11   YEARS_BEGINEXPLUATATION_AVG     691056     48.320357  float64
39  YEARS_BEGINEXPLUATATION_MEDI     691056     48.320357  float64
25  YEARS_BEGINEXPLUATATION_MODE     691056     48.320357  float64
53                TOTALAREA_MODE     683808     47.813559  float64
55           EMERGENCYSTATE_MODE     671472     46.950995   object
97              DAYS_TERMINATION     577560     40.384434  float64
95     DAYS_LAST_DUE_1ST_VERSION     577560     40.384434  float64
96                 DAYS_LAST_DUE     577560     40.384434  float64
93            DAYS_FIRST_DRAWING     577560     40.384434  float64
94                DAYS_FIRST_DUE     577560     40.384434  float64
98     NFLAG_INSURED_ON_APPROVAL     577560     40.384434  float64
4                OCCUPATION_TYPE     462731     32.355304   object
72             AMT_GOODS_PRICE_y     335979     23.492489  float64
69                 AMT_ANNUITY_y     323672     22.631952  float64
90                   CNT_PAYMENT     323667     22.631603  float64
8                   EXT_SOURCE_3     245619     17.174292  float64
66    AMT_REQ_CREDIT_BUREAU_YEAR     165812     11.593988  float64
65     AMT_REQ_CREDIT_BUREAU_QRT     165812     11.593988  float64
64     AMT_REQ_CREDIT_BUREAU_MON     165812     11.593988  float64
63    AMT_REQ_CREDIT_BUREAU_WEEK     165812     11.593988  float64
62     AMT_REQ_CREDIT_BUREAU_DAY     165812     11.593988  float64
61    AMT_REQ_CREDIT_BUREAU_HOUR     165812     11.593988  float64

#### These columns don't have enough missing values to drop them, but they have enough missing values that dropping those rows would reduce the amount of data available.
#### Since the missing values are not changing the type of the columns, meaning they are not affecting any statistical operations on the columns, it is safe to leave these columns as they are.
#### However, care should be taken to take this into account when performing any operation involving these columns.

### 4. Outlier detection & handling

#### 4.1. Outlier detection

In [91]:
cols

Index(['SK_ID_CURR', 'TARGET', 'NAME_CONTRACT_TYPE_x', 'CODE_GENDER',
       'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN', 'AMT_INCOME_TOTAL',
       'AMT_CREDIT_x', 'AMT_ANNUITY_x',
       ...
       'NAME_SELLER_INDUSTRY', 'CNT_PAYMENT', 'NAME_YIELD_GROUP',
       'PRODUCT_COMBINATION', 'DAYS_FIRST_DRAWING', 'DAYS_FIRST_DUE',
       'DAYS_LAST_DUE_1ST_VERSION', 'DAYS_LAST_DUE', 'DAYS_TERMINATION',
       'NFLAG_INSURED_ON_APPROVAL'],
      dtype='object', length=154)

In [93]:
# Collect all numeric columns into one dataframe
all_apps_numeric = pd.DataFrame()

# Loop over columns and check the type
for col in all_apps.columns:
    if ('_ID_' in col) or ('FLAG_' in col):
        # Ignore ids and flags
        continue
    if (np.issubdtype(all_apps[col], np.number)):
        all_apps_numeric[col] = all_apps[col]

In [94]:
all_apps_numeric.head()

TARGET  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT_x  AMT_ANNUITY_x  \
0       1             0          202500.0      406597.5        24700.5   
1       0             0          270000.0     1293502.5        35698.5   
2       0             0          270000.0     1293502.5        35698.5   
3       0             0          270000.0     1293502.5        35698.5   
4       0             0           67500.0      135000.0         6750.0   

   AMT_GOODS_PRICE_x  REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  \
0           351000.0                    0.018801       -9461           -637   
1          1129500.0                    0.003541      -16765          -1188   
2          1129500.0                    0.003541      -16765          -1188   
3          1129500.0                    0.003541      -16765          -1188   
4           135000.0                    0.010032      -19046           -225   

   DAYS_REGISTRATION  ...  AMT_GOODS_PRICE_y  HOUR_APPR_PROCESS_START_y  \
0            -3648.0  ...           179055.0                        9.0   
1            -1186.0  ...           900000.0                       12.0   
2            -1186.0  ...           337500.0                       17.0   
3            -1186.0  ...            68809.5                       15.0   
4            -4260.0  ...            24282.0                        5.0   

   DAYS_DECISION  SELLERPLACE_AREA  CNT_PAYMENT  DAYS_FIRST_DRAWING  \
0         -606.0             500.0         24.0            365243.0   
1         -746.0              -1.0         12.0            365243.0   
2         -828.0            1400.0          6.0            365243.0   
3        -2341.0             200.0         12.0            365243.0   
4         -815.0              30.0          4.0            365243.0   

   DAYS_FIRST_DUE  DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE  DAYS_TERMINATION  
0          -565.0                      125.0          -25.0             -17.0  
1          -716.0                     -386.0         -536.0            -527.0  
2          -797.0                     -647.0         -647.0            -639.0  
3         -2310.0                    -1980.0        -1980.0           -1976.0  
4          -784.0                     -694.0         -724.0            -714.0  

[5 rows x 59 columns]

In [97]:
all_apps_outliers = pd.DataFrame(columns={'name', 'num outliers', 'outlier percent'})

# Get number of columns
sz = all_apps_numeric.shape[0]

# Loop over columns to find the first and third quartiles, the IQR, and any outliers
for col in all_apps_numeric.columns:
    descr = all_apps_numeric[col].describe()

    first = descr['25%']
    third = descr['75%']
    iqr = third - first
    lower = first - (1.5 * iqr)
    upper = third + (1.5 * iqr)

    # Filter outliers
    outliers = all_apps_numeric[(all_apps_numeric[col] < lower) | (all_apps_numeric[col] > upper)]

    num = outliers.shape[0]

    if (num == 0):
        continue

    percent = num / sz * 100

    # record in DataFrame
    all_apps_outliers = all_apps_outliers.append({'name': col, 'num outliers': num, 'outlier percent': percent}, ignore_index=True)

In [99]:
all_apps_outliers.sort_values(by='outlier percent', ascending=False, inplace=True)

In [100]:
all_apps_outliers

outlier percent                          name num outliers
10        24.934796          REGION_RATING_CLIENT       350287
7         24.396076                 DAYS_EMPLOYED       342719
11        24.017449   REGION_RATING_CLIENT_W_CITY       337400
17        22.201191        REG_CITY_NOT_WORK_CITY       311885
41        21.266191     AMT_REQ_CREDIT_BUREAU_QRT       298750
18        17.396990       LIVE_CITY_NOT_WORK_CITY       244395
40        16.328377     AMT_REQ_CREDIT_BUREAU_MON       229383
49        16.059160              SELLERPLACE_AREA       225601
55        13.732870              DAYS_TERMINATION       192921
54        12.793954                 DAYS_LAST_DUE       179731
44        12.412978               AMT_APPLICATION       174379
33        12.172376      DEF_30_CNT_SOCIAL_CIRCLE       170999
45        10.790839                  AMT_CREDIT_y       151591
35         8.979422      DEF_60_CNT_SOCIAL_CIRCLE       126144
46         8.782599             AMT_GOODS_PRICE_y       123379
0          8.663010                        TARGET       121699
32         7.574110      OBS_30_CNT_SOCIAL_CIRCLE       106402
34         7.432454      OBS_60_CNT_SOCIAL_CIRCLE       104412
16         7.354222        REG_CITY_NOT_LIVE_CITY       103313
53         5.616979     DAYS_LAST_DUE_1ST_VERSION        78908
43         5.024943                 AMT_ANNUITY_y        70591
2          4.551570              AMT_INCOME_TOTAL        63941
14         4.389199    REG_REGION_NOT_WORK_REGION        61660
26         4.312890               LIVINGAREA_MODE        60588
30         4.269397               LIVINGAREA_MEDI        59977
22         4.180915                LIVINGAREA_AVG        58734
31         4.109518                TOTALAREA_MODE        57731
5          4.047801             AMT_GOODS_PRICE_x        56864
51         3.798871            DAYS_FIRST_DRAWING        53367
15         3.595072   LIVE_REGION_NOT_WORK_REGION        50504
50         3.379242                   CNT_PAYMENT        47472
39         3.226837    AMT_REQ_CREDIT_BUREAU_WEEK        45331
52         2.410145                DAYS_FIRST_DUE        33858
6          2.398613    REGION_POPULATION_RELATIVE        33696
4          2.112881                 AMT_ANNUITY_x        29682
3          1.693465                  AMT_CREDIT_x        23790
19         1.602421   YEARS_BEGINEXPLUATATION_AVG        22511
27         1.595943  YEARS_BEGINEXPLUATATION_MEDI        22420
29         1.591814                FLOORSMAX_MEDI        22362
21         1.544762                 FLOORSMAX_AVG        21701
25         1.513654                FLOORSMAX_MODE        21264
23         1.462331  YEARS_BEGINEXPLUATATION_MODE        20543
1          1.396201                  CNT_CHILDREN        19614
42         1.322241    AMT_REQ_CREDIT_BUREAU_YEAR        18575
9          1.311706               CNT_FAM_MEMBERS        18427
28         1.242373                ENTRANCES_MEDI        17453
20         1.238244                 ENTRANCES_AVG        17395
24         1.230200                ENTRANCES_MODE        17282
13         1.207564    REG_REGION_NOT_LIVE_REGION        16964
12         0.673756     HOUR_APPR_PROCESS_START_x         9465
37         0.528256    AMT_REQ_CREDIT_BUREAU_HOUR         7421
38         0.505477     AMT_REQ_CREDIT_BUREAU_DAY         7101
48         0.472234                 DAYS_DECISION         6634
8          0.219033             DAYS_REGISTRATION         3077
47         0.098020     HOUR_APPR_PROCESS_START_y         1377
36         0.041856        DAYS_LAST_PHONE_CHANGE          588

In [102]:
all_apps_outliers.shape

(56, 3)

### 4.1.1. REGION_RATING_CLIENT

In [127]:
# Writing outliers to file for ease in analysis, since there are 56 such columns
def write_outliers_to_file(cols):
    with open('outliers_info.txt', 'w') as f:
        for col in cols:
            res = ''

            # print(col)
            res = res + 'Column: ' + col + '\n'

            res = res + str(all_apps[col].value_counts())

            res = res + '\n\n'

            f.write(res)


    f.close()

    # print(res)



In [128]:
write_outliers_to_file(list(all_apps_outliers['name'].values))

In [101]:
all_apps['REGION_RATING_CLIENT'].value_counts()

2    1054525
3     217268
1     133019
Name: REGION_RATING_CLIENT, dtype: int64

### 4.1.1. Wrongly detected outliers
#### These are not outliers. These are categorical variables where one option happens to be in more rows than the others. These should be converted into categorical variables
* REGION_RATING_CLIENT
* REGION_RATING_CLIENT_W_CITY
* REG_CITY_NOT_WORK_CITY
* LIVE_CITY_NOT_WORK_CITY
* REG_CITY_NOT_LIVE_CITY
* REG_REGION_NOT_WORK_REGION
* LIVE_REGION_NOT_WORK_REGION
* REG_REGION_NOT_LIVE_REGION

#### Unsure
* AMT_REQ_CREDIT_BUREAU_WEEK

#### These could be binned
* CNT_CHILDREN

#### Filter the loop
* TARGET

In [132]:
all_apps.HOUR_APPR_PROCESS_START_x.value_counts()

10    176225
11    170757
12    158055
13    140238
9     129972
14    125363
15    110904
16     90612
8      73432
17     64711
7      43272
18     38650
6      27154
5      16304
19     15537
4       9207
3       5162
20      4954
21      1803
2       1129
22       602
1        387
0        194
23       188
Name: HOUR_APPR_PROCESS_START_x, dtype: int64

In [131]:
all_apps.WEEKDAY_APPR_PROCESS_START_x.value_counts()

TUESDAY      247118
WEDNESDAY    234708
MONDAY       233130
FRIDAY       228800
THURSDAY     228784
SATURDAY     156034
SUNDAY        76238
Name: WEEKDAY_APPR_PROCESS_START_x, dtype: int64

In [134]:
all_apps.CNT_PAYMENT.value_counts()

12.0    273425
6.0     161664
10.0    121695
0.0     121365
24.0    116077
18.0     65157
36.0     60560
60.0     45451
48.0     40509
8.0      25926
4.0      23186
30.0     14521
14.0      7138
42.0      6201
16.0      4751
5.0       3468
54.0      1841
20.0      1600
7.0       1275
9.0       1113
3.0        936
15.0       835
11.0       576
72.0       121
13.0        48
84.0        45
17.0        37
22.0        32
23.0        24
26.0        11
66.0        10
35.0         9
28.0         8
29.0         6
32.0         5
19.0         5
34.0         3
59.0         3
21.0         3
41.0         3
39.0         2
47.0         2
45.0         2
38.0         1
40.0         1
53.0         1
33.0         1
44.0         1
Name: CNT_PAYMENT, dtype: int64

## 5. Imbalance

In [136]:
# Checking for imbalance in the data
all_apps.TARGET.value_counts(normalize=True)

0    0.91337
1    0.08663
Name: TARGET, dtype: float64

#### Positive cases of payment difficulty is ~8.7% of the data. This counts as Moderate data imbalance according to <a href="https://developers.google.com/machine-learning/data-prep/construct/sampling-splitting/imbalanced-data">this link</a>.
#### This imbalance would need to be corrected when moving to the data analysis step after EDA.